### ISSUE TITLE
train custom dataset

### ISSUE THREAD

**Issue by bitwangdan** [zylo117/Yet-Another-EfficientDet-Pytorch#66]
thanks for sharing the code, i trained on my own dataset( six classes), and i changed the coco.yaml, after 10 epochs,  the eval result is 0,  in training,  the loss is normal

**Comment by bitwangdan** [comment]
@zylo117  is there anything wrong in my settings?

**Comment by samson-wang** [comment]
Me too. Very weird predictions. Normal loss decrease.

**Comment by bitwangdan** [comment]
> Me too. Very weird predictions. Normal loss decrease.

my prediction box is a  negative number in "val2017_bbox_results.json"

**Comment by zylo117** [comment]
I might release a tutorial about how to train on a custom dataset. stay tune

**Comment by bitwangdan** [comment]
> I might release a tutorial about how to train on a custom dataset. stay tune

thanks

**Comment by bitwangdan** [comment]
![image](https://user-images.githubusercontent.com/18326564/79184992-48a8f980-7e48-11ea-867d-7e1a0428e137.png)
@zylo117  my prediction result is like this,  my custom data should be right, i trained in other model works well

**Comment by bitwangdan** [comment]
![image](https://user-images.githubusercontent.com/18326564/79185611-05e82100-7e4a-11ea-8814-116891464414.png)
hi，This is the result of my training according to the training course you just updated (dataset_shape), what is your training result?

**Comment by bitwangdan** [comment]
@zylo117

**Comment by zylo117** [comment]
> @zylo117

sorry, it's still being tuned. I will update when it's done

**Comment by bitwangdan** [comment]
> > @zylo117
> 
> sorry, it's still being tuned. I will update when it's done

good

**Comment by zylo117** [comment]
it's a bug, so sorry for the trouble. please pull the latest code and try again

**Comment by zylo117** [comment]
> ![image](https://user-images.githubusercontent.com/18326564/79185611-05e82100-7e4a-11ea-8814-116891464414.png)
> hi，This is the result of my training according to the training course you just updated (dataset_shape), what is your training result?

This is the result after 6 epochs, the bug should be fixed by now. apologize again.
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.726
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.952
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.834
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.757
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.638
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.444
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.776
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.776
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.802
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.702

**Comment by samson-wang** [comment]
Start working!

**Comment by bitwangdan** [comment]
@samson-wang hi, Can you train normally? Is it right?

**Comment by zizizihua** [comment]
![39a389197a784d31aaf1b046d11292a1](https://user-images.githubusercontent.com/51010057/79226653-4fa82a00-7e91-11ea-8942-e148178ef35a.jpg)
After 50 epochs, the trainning loss is less than 0.05,  but get bad prediction

**Comment by zylo117** [comment]
> ![39a389197a784d31aaf1b046d11292a1](https://user-images.githubusercontent.com/51010057/79226653-4fa82a00-7e91-11ea-8942-e148178ef35a.jpg)
> After 50 epochs, the trainning loss is less than 0.05, but get bad prediction. What might be the reason？
> The picture is written by training process with '--debug=True'.

it is the latest code? the bug hasn't been fixed till 9 hrs ago.

pls share tensorboard loss graph, log, all of your hyperparammeters, etc.

**Comment by zizizihua** [comment]
it's the latest code. All hyperparammeters are same with default,  only changed num_classes.

**Comment by zylo117** [comment]
> it's the latest code. All hyperparammeters are same with default, only changed num_classes.

what is the dataset, coco?
did you train from scratch without pretrained?

you need to provide more info or the minimal code to replicate this issue.

**Comment by samson-wang** [comment]
@bitwangdan Yes, I managed to train on my own dataset. However, not as good as expected. Still need some tuning. Maybe the hyperparameters not right for my dataset. But the config file support too few params.

@zizizihua It seems that you use a too small learning rate.

**Comment by bitwangdan** [comment]
> it's the latest code. All hyperparammeters are same with default, only changed num_classes.

Is it the result of previous epochs?After a dozen epochs, the effect is better, but it seems to converge a bit slowly

**Comment by zylo117** [comment]
> @bitwangdan Yes, I managed to train on my own dataset. However, not as good as expected. Still need some tuning. Maybe the hyperparameters not right for my dataset. But the config file support too few params.
> 
> @zizizihua It seems that you use a too small learning rate.

Did you try running the tutorial about how to train on a custom dataset? It works, so maybe you can find out what actually went wrong by comparing it with yours.

**Comment by zizizihua** [comment]
Thanks. I ran the tutorial and it did work. I' m running my dataset once again. Expecting good results.

**Comment by zheng-2019** [comment]
![image](https://user-images.githubusercontent.com/26559584/79293390-3392a000-7f06-11ea-9531-8afd38fc962c.png)
i train custom data content 1 class，but the loss can't turn small

**Comment by zylo117** [comment]
> ![image](https://user-images.githubusercontent.com/26559584/79293390-3392a000-7f06-11ea-9531-8afd38fc962c.png)
> i train custom data content 1 class，but the loss can't turn small
> @zylo117 and resume training is not normal,when i resume train,the loss beacma very large

it's expoding gradients. try 100 times smaller lr

**Comment by samson-wang** [comment]
I've tried many other one-stage models, none of them worked as good as they claimed. I also tried bifpn and efficientnet in two-stage models, no difference between traditional se-resnet. Maybe more work should be done.

However, according to Kaiming's paper, a resnet50 can get 40+ or 41+ map on coco. 

So 模型无好坏，炼丹有高低

**Comment by zheng-2019** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"WANG Chenxi"<notifications@github.com&gt;;
发送时间:&nbsp;2020年4月15日(星期三) 中午11:11
收件人:&nbsp;"zylo117/Yet-Another-EfficientDet-Pytorch"<Yet-Another-EfficientDet-Pytorch@noreply.github.com&gt;;
抄送:&nbsp;"郑江红"<526078586@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [zylo117/Yet-Another-EfficientDet-Pytorch] train custom dataset (#66)





 
I've tried many other one-stage models, none of them worked as good as they claimed. I also tried bifpn and efficientnet in two-stage models, no difference between traditional se-resnet. Maybe more work should be done.
 
However, according to Kaiming's paper, a resnet50 can get 40+ or 41+ map on coco.
 
So 模型无好坏，炼丹有高低
 
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.




ths !!! i try again

**Comment by bitwangdan** [comment]
@zylo117 
![image](https://user-images.githubusercontent.com/18326564/79296425-f7fbd400-7f0d-11ea-9272-12e7771cb523.png)
这是我50个epoch后的结果，map:42%，the loss is very low，在其他efficientdet的project中，map is 60%, Could it be the problem of anchor size?

**Comment by zylo117** [comment]
> @zylo117
> ![image](https://user-images.githubusercontent.com/18326564/79296425-f7fbd400-7f0d-11ea-9272-12e7771cb523.png)
> 这是我50个epoch后的结果，map:42%，the loss is very low，在其他efficientdet的project中，map is 60%, Could it be the problem of anchor size?

you need to check on loss graph, it might not converge yet.

**Comment by bitwangdan** [comment]
@zylo117 hi, here 
![image](https://user-images.githubusercontent.com/18326564/79297914-25e31780-7f12-11ea-9156-e50df5ce12ad.png)
![image](https://user-images.githubusercontent.com/18326564/79297944-398e7e00-7f12-11ea-9f1d-79c967b75541.png)
![image](https://user-images.githubusercontent.com/18326564/79297958-414e2280-7f12-11ea-9b87-d788bf148acd.png)
![image](https://user-images.githubusercontent.com/18326564/79297968-490dc700-7f12-11ea-86aa-f230c49bfdaf.png)
It looks like it's completely converged。
python train.py -c 0 --batch_size 8 --lr 1e-5 --num_epochs 100    --load_weights ./weights/efficientdet-d0.pth
I trained with these parameters

**Comment by zylo117** [comment]
no. it's just overfitting...

**Comment by bitwangdan** [comment]
> no. it's just overfitting...
hi，Can you give me some advice? My idea is to narrow the learning rate and adjust the rules of learning rate change

**Comment by bitwangdan** [comment]
@zylo117 hi, How do you see that the model has been overfitting? The convergence seems normal

**Comment by cendelian** [comment]
hi I also have the same problem.
I just train one class(person), after 70 epoch. I test one img, and prediction is so bad.
my tain command is :
`python person_train.py --lr 1e-5 --batch_size 12 --load_weights ./weights/efficientdet-d0.pth`


and one img's test result is as below.
there are so many box.
and i think the loss is enough but not overfitting.

could you help me?

![image](https://user-images.githubusercontent.com/50360389/79342098-4d140600-7f5f-11ea-8af3-fb4f0c92e8f3.png)

**Comment by cendelian** [comment]
I add more one test result
![image](https://user-images.githubusercontent.com/50360389/79342598-e6dbb300-7f5f-11ea-8221-33c50db50676.png)

I think the anchor ratio is wrong, 
but why so many bbox?

anything help is great.

thanks

**Comment by zylo117** [comment]
1. make sure the anchor config is consistent among training,evaluation,inference

2. make sure image size is the same as training while inferring, which mean force input size must be set to None

3. it's normal that there are false positive boxes, especially in early stage.

4. loss being low doesn't mean no overfitting.

5. person is a difficult class to train while not many folks realize this.

6. add `--debug True`, then check out test/ folder, there are annotated images during training, it helps visualizing the performance.

7. the right way to judge the model's performance is to run coco_eval

**Comment by bitwangdan** [comment]
@zylo117 
hi, Thanks for your reply，this is the result of my training on my own data with the efficientdet-d1 model, I put the learning rate to 10 times, 1e-4,the map on my dataset is 57%  after 15 epochs, close to yolo v3 and other efficientdet projects (63%),I will continue training.this is my loss result, does this look normal?
![image](https://user-images.githubusercontent.com/18326564/79346575-21941a00-7f65-11ea-8e3a-65d83c2c1449.png)
![image](https://user-images.githubusercontent.com/18326564/79346723-46888d00-7f65-11ea-8b1d-fa58ebd5d202.png)
![image](https://user-images.githubusercontent.com/18326564/79346756-51432200-7f65-11ea-9cd9-42ca8851ff0b.png)

**Comment by zylo117** [comment]
> @zylo117
> hi, Thanks for your reply，this is the result of my training on my own data with the efficientdet-d1 model, I put the learning rate to 10 times, 1e-4,the map on my dataset is 57% after 15 epochs, close to yolo v3 and other efficientdet projects (63%),I will continue training.this is my loss result, does this look normal?
> ![image](https://user-images.githubusercontent.com/18326564/79346575-21941a00-7f65-11ea-8e3a-65d83c2c1449.png)
> ![image](https://user-images.githubusercontent.com/18326564/79346723-46888d00-7f65-11ea-8b1d-fa58ebd5d202.png)
> ![image](https://user-images.githubusercontent.com/18326564/79346756-51432200-7f65-11ea-9cd9-42ca8851ff0b.png)

Good job! It is effdet D3 on offiical coco dataset, or the a slim coco dataset?

Anyway, congratulations! 

Keep training. It looks like still a long way to go.

**Comment by zizizihua** [comment]
@zylo117 Thanks for your work. Finally it got map 81% by d4 on my dataset. It's a nice performance.

**Comment by bitwangdan** [comment]
> @zylo117 Thanks for your work. Finally it got map 81% by d4 on my dataset. It's a nice performance.

hi, Have you adjusted the lr?

**Comment by cendelian** [comment]
> @zylo117 Thanks for your work. Finally it got map 81% by d4 on my dataset. It's a nice performance.

have you change the anchor ratio? my prediction is so bad

**Comment by zizizihua** [comment]
@bitwangdan I used default lr 1e-4. The lowest val loss is 0.08.

**Comment by bitwangdan** [comment]
> @bitwangdan I used default lr 1e-4. The lowest val loss is 0.08.
okay,thanks

**Comment by zizizihua** [comment]
@cendelian no, my dataset is similar to coco, just less class num. Did you set score thresh while inferring image. I found that my eval_json_result contains lots of bbox which score is less than 0.1.

**Comment by cendelian** [comment]
> 1. make sure the anchor config is consistent among training,evaluation,inference
the default anchor ratio come from coco. but i think it's not resonable to get too bad bbox(too long and shin)
> 2. make sure image size is the same as training while inferring, which mean force input size must be set to None
Yes , they are same
> 3. it's normal that there are false positive boxes, especially in early stage.
> 4. loss being low doesn't mean no overfitting.
I have test the 20 epoch. 40 epoch. 70 epoch. they are all bad prediction.
> 5. person is a difficult class to train while not many folks realize this.
I use the same data to train yolov3, the result is good.
> 6. add `--debug True`, then check out test/ folder, there are annotated images during training, it helps visualizing the performance.
I did it. after 3 epoch, the prediction is so bad too.
> 7. the right way to judge the model's performance is to run coco_eval
I run the coco_eval for my person data. the map is too low....

I will continue to try.

and if you have any suggestion. it great.
thanks again.

**Comment by zheng-2019** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"zizizihua"<notifications@github.com&gt;;
发送时间:&nbsp;2020年4月16日(星期四) 上午10:18
收件人:&nbsp;"zylo117/Yet-Another-EfficientDet-Pytorch"<Yet-Another-EfficientDet-Pytorch@noreply.github.com&gt;;
抄送:&nbsp;"郑江红"<526078586@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [zylo117/Yet-Another-EfficientDet-Pytorch] train custom dataset (#66)





 
@bitwangdan I used default lr 1e-4. The lowest val loss is 0.08.
 
—
You are receiving this because you commented.
 directly, Reply to this email directly, view it on GitHub, or unsubscribe.




i train my dataset, only contern 1 class,finally,the eval f1 = 0.412

**Comment by zheng-2019** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"zizizihua"<notifications@github.com&gt;;
发送时间:&nbsp;2020年4月16日(星期四) 上午10:22
收件人:&nbsp;"zylo117/Yet-Another-EfficientDet-Pytorch"<Yet-Another-EfficientDet-Pytorch@noreply.github.com&gt;;
抄送:&nbsp;"郑江红"<526078586@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [zylo117/Yet-Another-EfficientDet-Pytorch] train custom dataset (#66)





 
@cendelian no, my dataset is similar to coco, just less class num. Did you set score thresh while inferring image. I found that my eval_json_result contains lots of bbox which score is less than 0.1.
 
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.




i alse train my dataset, only contern 1 class,fanally,the eval f1 = 0.412

**Comment by Cli98** [comment]
In case of your need to compute anchors, you may have a look here
https://github.com/Cli98/anchor_computation_tool

**Comment by mytk2012** [comment]
> hi I also have the same problem.
> I just train one class(person), after 70 epoch. I test one img, and prediction is so bad.
> my tain command is :
> `python person_train.py --lr 1e-5 --batch_size 12 --load_weights ./weights/efficientdet-d0.pth`
> 
> and one img's test result is as below.
> there are so many box.
> and i think the loss is enough low but not overfitting.
> 
> could you help me?
> 
> ![image](https://user-images.githubusercontent.com/50360389/79342098-4d140600-7f5f-11ea-8af3-fb4f0c92e8f3.png)
> 
> by the way. the person data come from voc2012..

How can you see your test results?
